### DISCUSSION TITLE
phpmailer help

### DISCUSSION THREAD

**Discussion by metalchipper** [PHPMailer/PHPMailer#2876] [General]
Hello.
About ten years ago I made a web site for a friend. But now the server does not support the 
mail-form-php that we used to email the customer’s information. So now I am trying to use phpmailer. It is a mess so I am hoping some one can help me clean it up. Thanks in advance!
Metalchipper

I am unsure what to fill in the “blanks” with.   Now ……

I uploaded Composer into the server and added the below line.

Added this on the server …..in the composer.json file
added: `"phpmailer/phpmailer": "^6.7.1"`

I have the `phpmailer.php` in the `/vender/phpmailer/phpmailer/phpmailer.php` on the server

I called the server, and they said I should use port 587

On my browser, upon “submit” I get this
 
Error 404 - Not Found
The document you are looking for may have been removed or re-named. Please contact the web site owner for further assistance.

From the program Visual Studio Code : this
Not Found
The requested URL was not found on this server.
________________________________________
Apache/2.4.54 (Win64) OpenSSL/1.1.1p PHP/8.2.0 Server at localhost Port 80

This is from the top and bottom of the form, from the html web page:
```
</p>
        <form method="post" form action="../_phpmailer/phpmailer/phpmailer.php" enctype="multipart/form-data">
          <input type="hidden" name="referer" value="3contact.html" />
          <input type="hidden" name="required" value="firstname,lastname,streetaddress,city,state,zip, phonenumber,email" />
          <input type="hidden" name="print_config" value="title,sort" />
          &nbsp;
          <table class="border" width="92%" border="0" cellpadding="0" cellspacing="0" bgcolor="#EAEAEA">
               
              <td width="4" rowspan="38" bgcolor="#FFFFFF">&nbsp;</td>
              <td width="120" nowrap bgcolor="#EAEAEA">&nbsp;</td>
              <td width="4" bgcolor="#EAEAEA">&nbsp;</td>
              <td width="120" nowrap bgcolor="#EAEAEA"><strong>Order Form</strong></td>
              <td width="23" bgcolor="#EAEAEA">&nbsp;</td>
              <td width="120" nowrap bgcolor="#EAEAEA"><SCRIPT LANGUAGE="JavaScript"> 
if (window.print) {
document.write('<form><input type=button name=print value="Print this page" onClick="window.print()"></form>');
}
[From the end of the form]

  <tr>
              <td nowrap bgcolor="#EAEAEA"><input type="reset" name="button2" id="button2" value="Reset Form"></td>
              <td bgcolor="#EAEAEA">&nbsp;</td>
              <td nowrap bgcolor="#EAEAEA">&nbsp;</td>
              <td bgcolor="#EAEAEA">&nbsp;</td>
              <td nowrap bgcolor="#EAEAEA"><input type="submit" name="button" id="button" value="Order Now"></td>
            </tr>
          </table>
        </form>
```
This is the phpmailer:  from github’s simple form example

```php
<?php
 
/**
 * PHPMailer simple contact form example.
 * If you want to accept and send uploads in your form, look at the send_file_upload example.
 */

//Import the PHPMailer class into the global namespace
use PHPMailer\PHPMailer\PHPMailer;


require vendor/autoload.php


 (array_key_exists('to', $_POST)) {
    $err = true;
    $msg = '';
    $email = '';
    //Apply some basic validation and filtering to the subject
    if (array_key_exists('required', $_POST)) {
        $subject = substr(strip_tags($_POST['required']), 0, 255);
    } else {
        $subject = 'No subject given';
    }
    //Apply some basic validation and filtering to the query
    if (array_key_exists('query', $_POST)) {
        //Limit length and strip HTML tags
        $query = substr(strip_tags($_POST['query']), 0, 16384);
    } else {
        $query = '';
        $msg = 'No query provided!';
        $err = true;
    }
    //Apply some basic validation and filtering to the name
    if (array_key_exists('name', $_POST)) {
        //Limit length and strip HTML tags
        $name = substr(strip_tags($_POST['name']), 0, 255);
    } else {
        $name = '';
    }
    //Validate to address
    //Never allow arbitrary input for the 'to' address as it will turn your form into a spam gateway!
    //Substitute appropriate addresses from your own domain, or simply use a single, fixed address
    if (array_key_exists('to', $_POST) && in_array($_POST['steve@xxemailxx.com'],  true)) {
        $to = $_POST['steve@xxemailxx.com'] . 'http://www.xxwebsitexx.com';
    } 
    //Make sure the address they provided is valid before trying to use it
    if (array_key_exists('email', $_POST) && PHPMailer::validateAddress($_POST['email'])) {
        $email = $_POST['email'];
    } else {
        $msg .= 'Error: invalid email address provided';
        $err = true;
    }
    if (!$err) {
        $mail = new PHPMailer();
        $mail->isSMTP();
        
        $mail->Host = 'mail.xxwebsitexxx.com';
        $mail->SMTPAuth = true;
        $mail->Username = 'steve@xxxemailxx.com';
        $mail->Password = 'xxxxxx';		Note: the server support suggested to add the password
        $mail->SMTPSecure = 'ssl';		And use ssl
        $mail->Port = 587;			And use this port #
        $mail->CharSet = PHPMailer::CHARSET_UTF8;
        //It's important not to use the submitter's address as the from address as it's forgery,
        //which will cause your messages to fail SPF checks.
        //Use an address in your own domain as the from address, put the submitter's address in a reply-to
        $mail->setFrom('3contact.html', (empty($name) ? 'referer' : $name));
        $mail->addAddress($to);
        $mail->addReplyTo($email, $name);
        $mail->Subject = 'Contact / Order: ' . $subject;
        $mail->Body = "Contact form submission\n\n" . $query;
        if (!$mail->send()) {
            $msg .= 'Mailer Error: ' . $mail->ErrorInfo;
        } else {
            $msg .= 'Message sent!';
        }
    }
} ?>

```

**Comment by Synchro** [comment]
This code doesn't make much sense:

```php
 if (array_key_exists('to', $_POST) && in_array($_POST['steve@xxemailxx.com'],  true)) {
        $to = $_POST['steve@xxemailxx.com'] . 'http://www.xxwebsitexx.com';
    }
```
That's not going to result in `$to` containing an email address. I suspect you simply want `$to = 'steve@xxemailxx.com'`, to send it to yourself.

`$mail->SMTPSecure = 'ssl';` won't work with `$mail->Port = 587;`; either change to `tls` mode, or change the port to `465` (but not both!).

`setFrom` expects an email address - so `3contact.html` won't work. It's quite likely you want to use your own address as the from address (as the comment above that says, and because you shouldn't forge the sender address), and you've already got the submitter's address in the `replyTo`, which is the right way to go.

**Reply by metalchipper** [reply]
Thanks Synchro,
I will make the changes and see what happens.  It has been a challenge for me and you are the first to offer some help. Most web designers would only do a complete new site, not wanting to fix this phpmailer problem.

**Comment by metalchipper** [comment]
Hello and thanks again,
I made the changes and am still getting the "404" error.
I tried several different..... $mail->Host = "addresses" I will call the server again to see what the correct address should be.

When I talked to the help at the server he said that the "vender" folder should be in the "public_html" folder
I stated that, would not be visible on the web? He said no. But I did not move the folder from where Composer installed it.

I have it linked from the html as        form action="../_phpmailer/phpmailer/phpmailer.php"
On the server the phpmailer.php is in        Vender/phpmailer/phpmailer/phpmailer.php"

My question is do I have the correct link from the html to the php?

**Reply by Synchro** [reply]
You have some funny naming going here. Typically you do not want your libraries to be inside your web root, but one level above, something like this (there will be other things here too, I've just shown the critical bits):

```
/home/username
    vendor/
        autoload.php
        composer/
        phpmailer/
            src/
                PHPMailer.php
    public_html/
        index.php
```
Your web page will be loaded from `index.php`, and would load composer's autoloader from `../vendor/autoload.php`. I note that your code above has `require vendor/autoload.php` – but this is not valid PHP syntax (missing quotes and a trailing semicolon) so this code will not even run.

A 404 error means it's not finding your web page, so I suspect you're not even getting as far as the PHP code you have here. Remember also that Linux file systems are usually case-sensitive, so `phpmailer.php` and `PHPMailer.php` are considered different.
